Intro to programming concepts via pseudocode. (#16)
Ben-Woolley/java-for-beginners: docs/lessons/lesson0.md

## Welcome to Java 101!
This course is intended to be an introduction to **programming that happens to be in Java**.
Many of the concepts you learn here are in the vast majority of programming languages out there; each one being a different tool for different jobs!

By the end of this course I hope you will have an awareness or understanding of:
* The fundamental concepts that programming languages feature
* Some basic object-oriented programming concepts

And the ability to:
* Solve problems by "translating English into Java"
* Write small, interactive, multi-file Java programs

## How do programs run on your computer?
Computers understand a fundamental set of instructions - these instructions are however very primitive and way too low-level for everyone to write programs in.
Programming languages sit on top of these primitive instructions and give us a more human-friendly set of tools.

The world now has more programming languages than it knows what to do with - built not only on top of machine code but also on each other.

Each has it's own specialty or programming opinions that makes it the best for certain jobs, for example:

> * **C** is a language written originally for building operating systems for computers. It exposes enough of the machine level to let clever people to magic, and dangerous people cause many more bugs.

Reviewer: ```suggestion
* **C** is a language written originally for building operating systems for computers. It exposes enough of the machine level to let clever people do magic, and dangerous people cause many more bugs.
```

* **R** specializes in statistical computing
* **Python** is a general-purpose language focusing on readability

These languages end up running on computers in one of two ways:

> * The program source code is **compiled** down to machine-understandable code.

Reviewer: ```suggestion
* The program source code is **compiled** (aka translated) to machine-understandable code **before it can be run**.
```

    - e.g. C, C++, Java*
* An **interpreter** runs the program that knows how to turn your source into machine-understandable code **while it runs**.
    - e.g. Python, Java*

> Java actually does both! Java source is compiled to Java Virtual Machine (JVM) bytecode, which the JVM installed on supported machines can then interpret!

Reviewer: ```suggestion
*Java actually does both! Java source is compiled to Java Virtual Machine (JVM) bytecode, which the JVM installed on supported machines can then interpret!
```

Author: I'd rather leave this as not part of the bullet list since its not one of the core ways in which programs are run

* Simulations
* Video games
* Development tools (like the one we'll be using)
* Android apps (In a sense, they use the same language, but Android doesn't use the JVM)

## Common pitfalls for beginners
As with learning most things, we all make the same mistakes and assumptions, here is a list of what I've noticed while teaching this course.

* **You can refer to the lesson material while doing exercises** - it's not an exam, treat it like you're solving any problem.
* **Lines that don't end in a bracket of some sort must end with a semicolon ;** - this is so Java knows the *expression* you're writing is done.
* **Java is *case sensitive*** - you must match the case and be consistent with your naming, otherwise your program will break!
* **Every open bracket, brace, and parenthesis must have a matching close** - this is how Java knows what context something is happening in. If they're in the wrong place you're likely going to run into problems.
* **If your program is in an unhappy state please ask for help!** - It's much easier for us to fix something as soon as it goes wrong, rather than after you finish writing and it won't compile.

> 

Reviewer: ```suggestion
* **Coding is both a science and an art** - There usually is more than one way of solving a problem. By all means consult with your fellow learners, but don't panic if their solutions are not the same as yours.
```
